### PR TITLE
grouping by digits is implemented

### PR DESCRIPTION
**Pull request by ani-hyan**
>

### CODE REVIEW

File: src/main/java/org/example/pubsub/blockingqueue_impl/Publisher.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+    public Publisher(BlockingQueue<String> messageQueue) {
+        this.messageQueue = messageQueue;
+    }
+
+    @Override
+    public void run() {
+        try {
+            for (int i = 0; i < 5; i++) {
+                synchronized (messageQueue) {
```
Comment (mchinarev) [inline src/main/java/org/example/pubsub/blockingqueue_impl/Publisher.java:17]: Do you need synchronized here? :)

> Memory consistency effects: As with other concurrent collections, actions in a thread prior to placing an object into a BlockingQueue [happen-before](https://docs.oracle.com/javase/6/docs/api/java/util/concurrent/package-summary.html#MemoryVisibility) actions subsequent to the access or removal of that element from the BlockingQueue in another thread.

https://docs.oracle.com/javase/6/docs/api/java/util/concurrent/BlockingQueue.html

Comment (mchinarev) [reply src/main/java/org/example/pubsub/blockingqueue_impl/Publisher.java:17]: Or I don't understand how it should work.. could you please add an example with the usage of these Publisher and Subscriber classes? You can just use one of the existing implementations of BlockingQueue, for example ArrayBlockingQueue

File: src/main/java/org/example/pubsub/parallelmerge_impl/ParallelMergeSort.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
+package org.example.pubsub.parallelmerge_impl;
+
+import java.util.Arrays;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.RecursiveAction;
+
+public class ParallelMergeSort<T extends Comparable<T>> {
```
Comment (mchinarev) [inline src/main/java/org/example/pubsub/parallelmerge_impl/ParallelMergeSort.java:7]: accepted 👍

File: src/main/java/org/example/pubsub/synchronized_impl/MessageBroker.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package org.example.pubsub.synchronized_impl;
+
+import java.util.LinkedList;
+import java.util.Queue;
+
+
+public class MessageBroker {
```
Comment (mchinarev) [inline src/main/java/org/example/pubsub/synchronized_impl/MessageBroker.java:7]: accepted 👍

File: src/main/java/org/example/pubsub/blockingqueue_impl/Publisher.java (hunk 1 above)
Comment (ani-hyan) [reply src/main/java/org/example/pubsub/blockingqueue_impl/Publisher.java:17]: This is one way we can use it
`
BlockingQueue<String> messageQueue = new ArrayBlockingQueue<>(5);

        Thread publisher2 = new Thread(new Publisher(messageQueue));
        Thread subscriber2 = new Thread(new Subscriber(messageQueue));

        publisher2.start();
        subscriber2.start();
        `        
        
I know BlockingQueue is thread-safe, but I wasn't getting the output I wanted without synchronization and there were cases where subscriber was receiving messages even when the queue was empty

Comment (mchinarev) [reply src/main/java/org/example/pubsub/blockingqueue_impl/Publisher.java:17]: Hmmm got it, maybe this is because of notify() call in Publisher

**Review by mchinarev — APPROVED** [review]

**Review by mchinarev — APPROVED** [review]

**Pull request merged**